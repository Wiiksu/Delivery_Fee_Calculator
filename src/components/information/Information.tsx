import { useState } from "react";
import { FaMinus, FaPlus } from "react-icons/fa";
import { items } from "../../helpers/InformationItems";

export default function Information() {
  const [expanded, setExpanded] = useState<string[]>([]);

  const toggleExpand = (title: string) => {
    if (expanded.includes(title)) {
      setExpanded(expanded.filter((otherTitles) => otherTitles !== title));
    } else {
      setExpanded([...expanded, title]);
    }
  };

  const collapseAllSubitems = () => {
    setExpanded([]);
  };

  return (
    <section className="bg-purple-400 text-purple-50 py-3 px-4 rounded shadow-lg h-fit w-[22rem] md:w-[27rem]">
      {items.map((item, index) => (
        <div key={index}>
          <div
            onClick={() => {
              toggleExpand(item.title);
              if (expanded.includes(item.title)) {
                collapseAllSubitems();
              }
            }}
            className="font-semibold flex items-center gap-2 cursor-pointer select-none w-fit"
          >
            {expanded.includes(item.title) ? (
              <FaMinus size={13} />
            ) : (
              <FaPlus size={13} />
            )}
            <span
              data-testid="mainItemTitle"
              className="translate-y-[-1px] text-lg"
            >
              {item.title}
            </span>
          </div>
          {expanded.includes(item.title) && (
            <ul className="translate-x-2">
              {item.subinformation.map((subItem, subIndex) => (
                <div className="flex flex-col" key={subIndex}>
                  <li
                    onClick={() => toggleExpand(subItem.title)}
                    className="font-semibold flex items-center gap-2 cursor-pointer select-none w-fit"
                  >
                    {expanded.includes(subItem.title) ? (
                      <FaMinus size={10} />
                    ) : (
                      <FaPlus size={10} />
                    )}
                    <span className="translate-y-[-1px] text-base">
                      {subItem.title}
                    </span>
                  </li>
                  {expanded.includes(subItem.title) && (
                    <p className="bg-purple-500 text-sm font-semibold py-2 px-4 rounded break-words my-1 translate-x-1 mr-5">
                      {subItem.info}
                    </p>
                  )}
                </div>
              ))}
            </ul>
          )}
        </div>
      ))}
    </section>
  );
}
