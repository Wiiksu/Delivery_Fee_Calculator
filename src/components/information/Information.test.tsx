import { expect, it, describe, beforeEach } from "vitest";
import { render, screen, fireEvent } from "@testing-library/react";
import Information from "./Information";

describe("<Information />", () => {
  beforeEach(() => {
    render(<Information />);
  });

  it("Should render correctly to the snapshot", () => {
    const { asFragment } = render(<Information />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("Should have main dropdown-title visible", () => {
    expect(screen.getByTestId("mainItemTitle")).toBeInTheDocument();
  });

  it("Should open the sub-dropdown-titles when clicking the main-title", () => {
    fireEvent.click(screen.getByTestId("mainItemTitle"));
    expect(screen.getByText("Cart Value")).toBeInTheDocument();
  });

  it("Should open the sub-information-content when clicking on the sub-information title", () => {
    fireEvent.click(screen.getByTestId("mainItemTitle"));
    fireEvent.click(screen.getByText("Cart Value"));
    expect(
      screen.getByText(
        "If the cart value is less than 10€, a small order surcharge is added to the delivery price. The surcharge is the difference between the cart value and 10€. For example if the cart value is 8.90€, the surcharge will be 1.10€."
      )
    ).toBeInTheDocument();
  });
});
