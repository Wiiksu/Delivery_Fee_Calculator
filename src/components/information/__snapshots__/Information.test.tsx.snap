// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Information /> > Should render correctly to the snapshot 1`] = `
<DocumentFragment>
  <section
    class="bg-purple-400 text-purple-50 py-3 px-4 rounded shadow-lg h-fit w-[22rem] md:w-[27rem]"
  >
    <div>
      <div
        class="font-semibold flex items-center gap-2 cursor-pointer select-none w-fit"
      >
        <svg
          fill="currentColor"
          height="13"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="13"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
          />
        </svg>
        <span
          class="translate-y-[-1px] text-lg"
          data-testid="mainItemTitle"
        >
          About this app
        </span>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
